server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2543,6 +2545,39 @@ public void forceMerge(final boolean flush, int maxNumSegments, boolean onlyExpu
         }
     }
 
+    @Override
+    public boolean forceMergeIsNoOp(int maxNumSegments, boolean onlyExpungeDeletes) throws IOException {
+        if (onlyExpungeDeletes && maxNumSegments >= 0) {
+            throw new IllegalArgumentException("only_expunge_deletes and max_num_segments are mutually exclusive");
+        }
+        try (var reader = DirectoryReader.open(indexWriter)) {
+            final var segmentCommitInfos = SegmentInfos.readCommit(reader.directory(), reader.getIndexCommit().getSegmentsFileName());
+            final var segmentsToMerge = new HashMap<SegmentCommitInfo, Boolean>();
+            for (int i = 0; i < segmentCommitInfos.size(); i++) {
+                final var segmentInfo = segmentCommitInfos.info(i);
+                if (onlyExpungeDeletes && segmentInfo.hasDeletions()) {
+                    return false;
+                }
+                segmentsToMerge.put(segmentInfo, Boolean.TRUE);
```

Comment:
The reason why I believe you're seeing [test failures](https://gradle-enterprise.elastic.co/s/i72voxhk2vqby/tests/task/:server:internalClusterTest/details/org.elasticsearch.index.engine.MergeWithLowDiskSpaceIT/testRelocationWhileForceMerging?expanded-stacktrace=WyIwIl0&top-execution=1) is that the `IndexWriter#forceMerge` also looks at the segments that result from already running merges, ie https://github.com/apache/lucene/blob/47c1a77fad692411555ac9a919ff947a0f53cc18/lucene/core/src/java/org/apache/lucene/index/IndexWriter.java#L2143 . That's my best guess, but I could be wrong.

To me it doesn't look like such a good idea

Reply:
Looking closer, I do believe it's got something to do with concurrently running merges. When force merge is invoked it also looks at and affects currently running merges so that they "cascade" into a force merge when they're done: https://github.com/apache/lucene/blob/bace2920b4487384e4195a6914c8074c498ae4b5/lucene/core/src/java/org/apache/lucene/index/IndexWriter.java#L4672 .

Reply:
Looking some more, maybe concurrent merges are not the thing we should worry about. Concurrent merges might produce some new segments based on the existing ones, but the existing ones that the merge is based on are visible in your implementation. Now, I'm not 100% sure that the check you've implemented will (always) show that there's some merging to be done when a merge was already triggered for those same segments (depending on the context between the 2 merge policy invocations)...